pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,302 @@
+		m := bytes.LastIndexByte(buffer, byte('\n'))
+
+		f.offset -= int64(n - m - 1)
+		n = m
+	}
+
+	buffer = buffer[:n]
+
+	reader := bytes.NewReader(buffer)
+
+	return &textReader{
+		scanner: bufio.NewScanner(reader),
+		logger:  f.logger,
+	}, err
+}
+
+// Next checks if there is another JSON object available.
+func (j jsonReader) Next() bool {
+	return j.decoder.More()
+}
+
+// Scan decodes the next JSON object into the provided structure.
+func (j jsonReader) Scan(i interface{}) error {
+	return j.decoder.Decode(&i)
+}
+
+// Next checks if there is another line available in the text file.
+func (f textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan scans the next line from the text file into the provided pointer to string.
+func (f textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
+		return nil
+	default:
+		return errNotPointer
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	return f.response.Close()
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice and updates the file offset.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	r, err := f.conn.RetrFrom(f.path, uint64(f.offset))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	r.Close()
+
+	copy(p, b)
+
+	readbytes := len(p)
+
+	if len(b) < len(p) {
+		f.offset += int64(len(b))
+		readbytes = len(b)
+
+		f.logger.Logf("Read %v bytes from %v", readbytes, f.path)
+
+		return len(b), io.EOF
+	}
+
+	f.logger.Logf("Read %v bytes from %v", readbytes, f.path)
+
+	f.offset += int64(readbytes)
+
+	return len(p), nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(resp)
+	if err != nil {
+		return 0, err
+	}
+
+	copy(p, b)
+
+	resp.Close()
+
+	f.logger.Logf("Read %v bytes from %v at offset of %v", len(p), f.path, off)
+
+	return len(b), nil
+}
+
+func (f *ftpFile) Check(whence int, offset, length int64) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 || offset > length {
+			return 0, ErrOutOfRange
+		}
+
+		return offset, nil
+
+	case io.SeekEnd:
+		if offset > 0 || offset < -length {
+			return 0, ErrOutOfRange
+		}
+
+		return length + offset, nil
+
+	case io.SeekCurrent:
+		if f.offset+offset >= length || f.offset+offset < 0 {
+			return 0, ErrOutOfRange
+		}
+
+		return f.offset + offset, nil
+	default:
+	}
+
+	return 0, ErrOutOfRange
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *ftpFile) Seek(offset int64, whence int) (int64, error) {
+	var (
+		err error
+		r   ftpResponse
+	)
+
+	r, err = f.conn.Retr(f.path)
```

Comment:
You can use short declaration syntax for `err` and only need to declare `r`:
```go
var r ftpResponse
r, err := f.conn.Retr(f.path)
```

Actually, it looks like `Retr` will return a `ftpResponse` already, because of how `ServerConn` is defined. So, you don’t even need to pre-declare `r` to override the type.

Reply:
Noted.